IncompatibleClassChangeError: class z.c.a.s.harvester.plugin.embedded.DeltaPlugin$SyntheticDeltaRead

Hi! Got exception in Spline initialization.

Version: _za.co.absa.spline.agent.spark:spark-3.2-spline-agent-bundle_2.12:1.0.1_

```
java.lang.RuntimeException: Plugin instantiation failure: class za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry$$anonfun$$nestedInanonfun$allPlugins$1$1.applyOrElse(AutoDiscoveryPluginRegistry.scala:54)
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry$$anonfun$$nestedInanonfun$allPlugins$1$1.applyOrElse(AutoDiscoveryPluginRegistry.scala:54)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at scala.util.Failure.recover(Try.scala:234)
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry.$anonfun$allPlugins$1(AutoDiscoveryPluginRegistry.scala:54)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry.<init>(AutoDiscoveryPluginRegistry.scala:51)
	at za.co.absa.spline.agent.SplineAgent$.create(SplineAgent.scala:66)
	at za.co.absa.spline.harvester.SparkLineageInitializer.createListener(SparkLineageInitializer.scala:162)
	at za.co.absa.spline.harvester.SparkLineageInitializer.$anonfun$createListener$6(SparkLineageInitializer.scala:139)
	at za.co.absa.spline.harvester.SparkLineageInitializer.withErrorHandling(SparkLineageInitializer.scala:176)
	at za.co.absa.spline.harvester.SparkLineageInitializer.createListener(SparkLineageInitializer.scala:138)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.<init>(SplineQueryExecutionListener.scala:37)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.spark.util.Utils$.$anonfun$loadExtensions$1(Utils.scala:2879)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.apache.spark.util.Utils$.loadExtensions(Utils.scala:2868)
	at org.apache.spark.sql.util.ExecutionListenerManager.$anonfun$new$1(QueryExecutionListener.scala:85)
	at org.apache.spark.sql.util.ExecutionListenerManager.$anonfun$new$1$adapted(QueryExecutionListener.scala:84)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.sql.util.ExecutionListenerManager.<init>(QueryExecutionListener.scala:84)
	at org.apache.spark.sql.internal.BaseSessionStateBuilder.$anonfun$listenerManager$2(BaseSessionStateBuilder.scala:330)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.internal.BaseSessionStateBuilder.listenerManager(BaseSessionStateBuilder.scala:330)
	at org.apache.spark.sql.internal.BaseSessionStateBuilder.build(BaseSessionStateBuilder.scala:358)
	at org.apache.spark.sql.SparkSession$.org$apache$spark$sql$SparkSession$$instantiateSessionState(SparkSession.scala:1148)
	at org.apache.spark.sql.SparkSession.$anonfun$sessionState$2(SparkSession.scala:158)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.SparkSession.sessionState$lzycompute(SparkSession.scala:156)
	at org.apache.spark.sql.SparkSession.sessionState(SparkSession.scala:153)
	at org.apache.spark.sql.DataFrameReader.<init>(DataFrameReader.scala:732)
	at org.apache.spark.sql.SparkSession.read(SparkSession.scala:658)
	at $line16.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:24)
	at $line16.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:28)
	at $line16.$read$$iw$$iw$$iw$$iw$$iw$$iw.<init>(<console>:30)
	at $line16.$read$$iw$$iw$$iw$$iw$$iw.<init>(<console>:32)
	at $line16.$read$$iw$$iw$$iw$$iw.<init>(<console>:34)
	at $line16.$read$$iw$$iw$$iw.<init>(<console>:36)
	at $line16.$read$$iw$$iw.<init>(<console>:38)
	at $line16.$read$$iw.<init>(<console>:40)
	at $line16.$read.<init>(<console>:42)
	at $line16.$read$.<init>(<console>:46)
	at $line16.$read$.<clinit>(<console>)
	at $line16.$eval$.$print$lzycompute(<console>:7)
	at $line16.$eval$.$print(<console>:6)
	at $line16.$eval.$print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:747)
	at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1020)
	at scala.tools.nsc.interpreter.IMain.$anonfun$interpret$1(IMain.scala:568)
	at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:36)
	at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:116)
	at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:41)
	at scala.tools.nsc.interpreter.IMain.loadAndRunReq$1(IMain.scala:567)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:594)
	at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:564)
	at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:865)
	at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:733)
	at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:435)
	at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:456)
	at org.apache.spark.repl.SparkILoop.process(SparkILoop.scala:239)
	at org.apache.spark.repl.Main$.doMain(Main.scala:78)
	at org.apache.spark.repl.Main$.main(Main.scala:58)
	at org.apache.spark.repl.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:955)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1043)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1052)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry.$anonfun$instantiatePlugin$1(AutoDiscoveryPluginRegistry.scala:73)
	at scala.util.Try$.apply(Try.scala:213)
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry.instantiatePlugin(AutoDiscoveryPluginRegistry.scala:63)
	at za.co.absa.spline.harvester.plugin.registry.AutoDiscoveryPluginRegistry.$anonfun$allPlugins$1(AutoDiscoveryPluginRegistry.scala:53)
	... 89 more
Caused by: java.lang.IncompatibleClassChangeError: class za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin$SyntheticDeltaRead has interface org.apache.spark.sql.catalyst.plans.logical.LeafNode as super class
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:473)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin.<init>(DeltaPlugin.scala:43)
	... 97 more
```

--- a/core/src/main/scala/za/co/absa/spline/harvester/LineageHarvester.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/LineageHarvester.scala
@@ -27,6 +27,7 @@ import za.co.absa.commons.CollectionImplicits._
 import za.co.absa.commons.graph.GraphImplicits._
 import za.co.absa.commons.lang.CachingConverter
 import za.co.absa.commons.reflect.ReflectionUtils
+import za.co.absa.commons.reflect.ReflectionUtils.extractValue
 import za.co.absa.commons.reflect.extractors.SafeTypeMatchingExtractor
 import za.co.absa.spline.agent.SplineAgent.FuncName
 import za.co.absa.spline.harvester.LineageHarvester._
@@ -37,6 +38,7 @@ import za.co.absa.spline.harvester.builder.write.{WriteCommand, WriteCommandExtr
 import za.co.absa.spline.harvester.converter.DataTypeConverter
 import za.co.absa.spline.harvester.iwd.IgnoredWriteDetectionStrategy
 import za.co.absa.spline.harvester.logging.ObjectStructureLogging
+import za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin.`_: MergeIntoCommand`
 import za.co.absa.spline.harvester.postprocessing.PostProcessor
 import za.co.absa.spline.producer.model._
 
@@ -205,6 +207,10 @@ class LineageHarvester(
           Seq(RddWrap(erdd.rdd))
         case lrdd: LogicalRDD =>
           Seq(RddWrap(lrdd.rdd))
+        case `_: MergeIntoCommand`(command) =>
+          val target = extractValue[LogicalPlan](command, "target")
+          val source = extractValue[LogicalPlan](command, "source")
+          Seq(PlanWrap(source), PlanWrap(target))
         case _ => plan.children.map(PlanWrap)
       }
     case RddWrap(rdd) =>

--- a/core/src/main/scala/za/co/absa/spline/harvester/builder/OperationNodeBuilderFactory.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/builder/OperationNodeBuilderFactory.scala
@@ -22,13 +22,13 @@ import za.co.absa.spline.harvester.IdGeneratorsBundle
 import za.co.absa.spline.harvester.LineageHarvester.{PlanOrRdd, PlanWrap, RddWrap}
 import za.co.absa.spline.harvester.builder.plan.read.ReadNodeBuilder
 import za.co.absa.spline.harvester.builder.plan.write.WriteNodeBuilder
-import za.co.absa.spline.harvester.builder.plan.{AggregateNodeBuilder, GenerateNodeBuilder, GenericPlanNodeBuilder, JoinNodeBuilder, MergeIntoNodeBuilder, ProjectNodeBuilder, UnionNodeBuilder, WindowNodeBuilder}
+import za.co.absa.spline.harvester.builder.plan._
 import za.co.absa.spline.harvester.builder.rdd.GenericRddNodeBuilder
 import za.co.absa.spline.harvester.builder.rdd.read.RddReadNodeBuilder
 import za.co.absa.spline.harvester.builder.read.ReadCommand
 import za.co.absa.spline.harvester.builder.write.WriteCommand
 import za.co.absa.spline.harvester.converter.{DataConverter, DataTypeConverter}
-import za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin
+import za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin.`_: MergeIntoCommand`
 import za.co.absa.spline.harvester.postprocessing.PostProcessor
 
 class OperationNodeBuilderFactory(
@@ -57,7 +57,7 @@ class OperationNodeBuilderFactory(
     case g: Generate => new GenerateNodeBuilder(g)(idGenerators, dataTypeConverter, dataConverter, postProcessor)
     case w: Window => new WindowNodeBuilder(w)(idGenerators, dataTypeConverter, dataConverter, postProcessor)
     case j: Join => new JoinNodeBuilder(j)(idGenerators, dataTypeConverter, dataConverter, postProcessor)
-    case m: DeltaPlugin.SyntheticDeltaMerge => new MergeIntoNodeBuilder(m)(idGenerators, dataTypeConverter, dataConverter, postProcessor)
+    case `_: MergeIntoCommand`(m) => new MergeIntoNodeBuilder(m)(idGenerators, dataTypeConverter, dataConverter, postProcessor)
     case _ => new GenericPlanNodeBuilder(lp)(idGenerators, dataTypeConverter, dataConverter, postProcessor)
   }
 

--- a/core/src/main/scala/za/co/absa/spline/harvester/builder/plan/MergeIntoNodeBuilder.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/builder/plan/MergeIntoNodeBuilder.scala
@@ -16,15 +16,16 @@
 
 package za.co.absa.spline.harvester.builder.plan
 
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import za.co.absa.commons.reflect.ReflectionUtils.extractValue
 import za.co.absa.spline.harvester.IdGeneratorsBundle
 import za.co.absa.spline.harvester.ModelConstants.CommonExtras
-import za.co.absa.spline.harvester.converter.{DataConverter, DataTypeConverter}
-import za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin
+import za.co.absa.spline.harvester.converter.{DataConverter, DataTypeConverter, OperationParamsConverter}
 import za.co.absa.spline.harvester.postprocessing.PostProcessor
-import za.co.absa.spline.producer.model.{AttrRef, Attribute, FunctionalExpression}
+import za.co.absa.spline.producer.model.{AttrRef, Attribute, DataOperation, FunctionalExpression}
 
 class MergeIntoNodeBuilder
-(override val logicalPlan: DeltaPlugin.SyntheticDeltaMerge)
+(override val logicalPlan: LogicalPlan)
   (idGenerators: IdGeneratorsBundle, dataTypeConverter: DataTypeConverter, dataConverter: DataConverter, postProcessor: PostProcessor)
   extends GenericPlanNodeBuilder(logicalPlan)(idGenerators, dataTypeConverter, dataConverter, postProcessor) {
 
@@ -45,5 +46,26 @@ class MergeIntoNodeBuilder
       extra = Map(CommonExtras.Synthetic -> true),
       name = attr1.name
     )
+  }
+
+  override def build(): DataOperation = {
+
+    val condition = extractValue[Any](logicalPlan, "condition").toString
+    val matchedClauses = extractValue[Seq[Any]](logicalPlan, "matchedClauses").map(_.toString)
+    val notMatchedClauses = extractValue[Seq[Any]](logicalPlan, "notMatchedClauses").map(_.toString)
+
+    val dop = DataOperation(
+      id = operationId,
+      name = logicalPlan.nodeName,
+      childIds = childIds,
+      output = outputAttributes.map(_.id),
+      params = Map(
+        "condition" -> condition,
+        "matchedClauses" -> matchedClauses,
+        "notMatchedClauses" -> notMatchedClauses),
+      extra = Map.empty
+    )
+
+    postProcessor.process(dop)
   }
 }

--- a/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
@@ -17,16 +17,15 @@
 package za.co.absa.spline.harvester.plugin.embedded
 
 import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
-import org.apache.spark.sql.catalyst.plans.logical.{BinaryNode, LeafNode, LogicalPlan}
-import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.catalyst.plans.logical.{BinaryNode, LogicalPlan}
 import org.apache.spark.sql.{SaveMode, SparkSession}
 import za.co.absa.commons.reflect.ReflectionUtils.extractValue
 import za.co.absa.commons.reflect.extractors.SafeTypeMatchingExtractor
 import za.co.absa.spline.agent.SplineAgent.FuncName
 import za.co.absa.spline.harvester.builder.SourceIdentifier
-import za.co.absa.spline.harvester.plugin.Plugin.{Precedence, ReadNodeInfo, WriteNodeInfo}
+import za.co.absa.spline.harvester.plugin.Plugin.{Precedence, WriteNodeInfo}
 import za.co.absa.spline.harvester.plugin.embedded.DeltaPlugin._
-import za.co.absa.spline.harvester.plugin.{Plugin, ReadNodeProcessing, WriteNodeProcessing}
+import za.co.absa.spline.harvester.plugin.{Plugin, WriteNodeProcessing}
 import za.co.absa.spline.harvester.qualifier.PathQualifier
 
 import javax.annotation.Priority
@@ -36,75 +35,55 @@ import scala.language.reflectiveCalls
 class DeltaPlugin(
   pathQualifier: PathQualifier,
   session: SparkSession
-) extends Plugin
-    with WriteNodeProcessing
-    with ReadNodeProcessing {
+) extends Plugin with WriteNodeProcessing {
 
   override val writeNodeProcessor: PartialFunction[(FuncName, LogicalPlan), WriteNodeInfo] = {
     case (_, `_: DeleteCommand`(command)) =>
-      val logicalRelation = extractLogicalRelation(command, "target")
-      val condition = extractValue[Option[Expression]](command, "condition")
-      val catalogTable = logicalRelation.catalogTable.get
-      val uri = catalogTable.location.toString
+      val target = extractValue[LogicalPlan](command, "target")
 
-      val sourceId = SourceIdentifier(Some("delta"), uri)
+      val condition = extractValue[Option[Expression]](command, "condition")
+      val qualifiedPath = extractPath(command, "tahoeFileIndex")
+
+      val sourceId = SourceIdentifier(Some("delta"), qualifiedPath)
       val params = Map("condition" -> condition.map(_.toString))
-      WriteNodeInfo(sourceId, SaveMode.Overwrite, SyntheticDeltaRead(logicalRelation.output, sourceId, params, logicalRelation), params)
+      WriteNodeInfo(sourceId, SaveMode.Overwrite, target, params)
 
     case (_, `_: UpdateCommand`(command)) =>
-      val logicalRelation = extractLogicalRelation(command, "target")
+      val target = extractValue[LogicalPlan](command, "target")
+
       val condition = extractValue[Option[Expression]](command, "condition")
       val updateExpressions = extractValue[Seq[Expression]](command, "updateExpressions")
-      val catalogTable = logicalRelation.catalogTable.get
-      val uri = catalogTable.location.toString
+      val qualifiedPath = extractPath(command, "tahoeFileIndex")
 
-      val sourceId = SourceIdentifier(Some("delta"), uri)
+      val sourceId = SourceIdentifier(Some("delta"), qualifiedPath)
       val params = Map("condition" -> condition.map(_.toString), "updateExpressions" -> updateExpressions.map(_.toString()))
-      WriteNodeInfo(sourceId, SaveMode.Overwrite, SyntheticDeltaRead(logicalRelation.output, sourceId, params, logicalRelation), params)
+      WriteNodeInfo(sourceId, SaveMode.Overwrite, target, params)
 
     case (_, `_: MergeIntoCommand`(command)) =>
-      val targetRelation = extractLogicalRelation(command, "target")
-      val targetCatalogTable = targetRelation.catalogTable.get
-      val targetUri = targetCatalogTable.location.toString
-      val targetId = SourceIdentifier(Some("delta"), targetUri)
-      val syntheticTargetRead = SyntheticDeltaRead(targetRelation.output, targetId, Map.empty[String, Any], targetRelation)
+      val target = extractValue[LogicalPlan](command, "target")
+      val source = extractValue[LogicalPlan](command, "source")
 
-      val sourceRelation = extractLogicalRelation(command, "source")
-      val sourceCatalogTable = sourceRelation.catalogTable.get
-      val sourceUri = sourceCatalogTable.location.toString
-      val sourceId = SourceIdentifier(Some("delta"), sourceUri)
-      val syntheticSourceRead = SyntheticDeltaRead(sourceRelation.output, sourceId, Map.empty[String, Any], sourceRelation)
-      val params = Map.empty[String, Any]
+      val qualifiedPath = extractPath(command, "targetFileIndex")
+      val targetId = SourceIdentifier(Some("delta"), qualifiedPath)
 
       val condition = extractValue[Expression](command, "condition").toString
       val matchedClauses = extractValue[Seq[Any]](command, "matchedClauses").map(_.toString)
       val notMatchedClauses = extractValue[Seq[Any]](command, "notMatchedClauses").map(_.toString)
 
-      val merge = SyntheticDeltaMerge(targetRelation.output, syntheticSourceRead, syntheticTargetRead, condition, matchedClauses, notMatchedClauses)
+      val merge = SyntheticDeltaMerge(target.output, source, target, condition, matchedClauses, notMatchedClauses)
 
-      WriteNodeInfo(targetId, SaveMode.Overwrite, merge, params)
+      WriteNodeInfo(targetId, SaveMode.Overwrite, merge, Map.empty[String, Any])
   }
 
-  private def extractLogicalRelation(o: AnyRef, fieldName: String): LogicalRelation = {
-    val logicalPlan = extractValue[LogicalPlan](o, fieldName)
-    extractValue[LogicalRelation](logicalPlan, "child")
+  private def extractPath(command: AnyRef, fieldName: String): String = {
+    val targetFileIndex = extractValue[AnyRef](command, fieldName)
+    val path = extractValue[org.apache.hadoop.fs.Path](targetFileIndex, "path")
+    pathQualifier.qualify(path.toString)
   }
 
-  override val readNodeProcessor: PartialFunction[LogicalPlan, ReadNodeInfo] = {
-    case SyntheticDeltaRead(output, sourceId, writeParams, logicalPlan) =>
-      val readParams = Map.empty[String, Any]
-      ReadNodeInfo(sourceId, readParams)
-  }
 }
 
 object DeltaPlugin {
-
-  case class SyntheticDeltaRead(
-    output: Seq[Attribute],
-    sourceId: SourceIdentifier,
-    writeParams: Map[String, Any],
-    logicalPlan: LogicalPlan
-  ) extends LeafNode
 
   case class SyntheticDeltaMerge(
     output: Seq[Attribute],

--- a/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
@@ -30,6 +30,7 @@ import za.co.absa.spline.harvester.qualifier.PathQualifier
 
 import javax.annotation.Priority
 import scala.language.reflectiveCalls
+import scala.util.Try
 
 @Priority(Precedence.Normal)
 class DeltaPlugin(
@@ -76,8 +77,13 @@ class DeltaPlugin(
   }
 
   private def extractPath(command: AnyRef, fieldName: String): String = {
-    val targetFileIndex = extractValue[AnyRef](command, fieldName)
-    val path = extractValue[org.apache.hadoop.fs.Path](targetFileIndex, "path")
+    val path = Try {
+      val targetFileIndex = extractValue[AnyRef](command, fieldName)
+      extractValue[org.apache.hadoop.fs.Path](targetFileIndex, "path")
+    }.getOrElse {
+      val deltaLog = extractValue[AnyRef](command, "deltaLog")
+      extractValue[AnyRef](deltaLog, "dataPath")
+    }
     pathQualifier.qualify(path.toString)
   }
 

--- a/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
@@ -16,8 +16,8 @@
 
 package za.co.absa.spline.harvester.plugin.embedded
 
-import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
-import org.apache.spark.sql.catalyst.plans.logical.{BinaryNode, LogicalPlan}
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.{SaveMode, SparkSession}
 import za.co.absa.commons.reflect.ReflectionUtils.extractValue
 import za.co.absa.commons.reflect.extractors.SafeTypeMatchingExtractor
@@ -61,19 +61,10 @@ class DeltaPlugin(
       WriteNodeInfo(sourceId, SaveMode.Overwrite, target, params)
 
     case (_, `_: MergeIntoCommand`(command)) =>
-      val target = extractValue[LogicalPlan](command, "target")
-      val source = extractValue[LogicalPlan](command, "source")
-
       val qualifiedPath = extractPath(command, "targetFileIndex")
       val targetId = SourceIdentifier(Some("delta"), qualifiedPath)
 
-      val condition = extractValue[Expression](command, "condition").toString
-      val matchedClauses = extractValue[Seq[Any]](command, "matchedClauses").map(_.toString)
-      val notMatchedClauses = extractValue[Seq[Any]](command, "notMatchedClauses").map(_.toString)
-
-      val merge = SyntheticDeltaMerge(target.output, source, target, condition, matchedClauses, notMatchedClauses)
-
-      WriteNodeInfo(targetId, SaveMode.Overwrite, merge, Map.empty[String, Any])
+      WriteNodeInfo(targetId, SaveMode.Overwrite, command, Map.empty[String, Any])
   }
 
   private def extractPath(command: AnyRef, fieldName: String): String = {
@@ -91,26 +82,13 @@ class DeltaPlugin(
 
 object DeltaPlugin {
 
-  case class SyntheticDeltaMerge(
-    output: Seq[Attribute],
-    left: LogicalPlan,
-    right: LogicalPlan,
-    condition: String,
-    matchedClauses: Seq[String],
-    notMatchedClauses: Seq[String]
-  ) extends BinaryNode {
-
-    protected def withNewChildrenInternal(newLeft: LogicalPlan, newRight: LogicalPlan): LogicalPlan =
-      throw new UnsupportedOperationException()
-  }
-
   private object `_: DeleteCommand` extends SafeTypeMatchingExtractor[AnyRef](
     "org.apache.spark.sql.delta.commands.DeleteCommand")
 
   private object `_: UpdateCommand` extends SafeTypeMatchingExtractor[AnyRef](
     "org.apache.spark.sql.delta.commands.UpdateCommand")
 
-  private object `_: MergeIntoCommand` extends SafeTypeMatchingExtractor[AnyRef](
+  object `_: MergeIntoCommand` extends SafeTypeMatchingExtractor[LogicalPlan](
     "org.apache.spark.sql.delta.commands.MergeIntoCommand")
 
 }

--- a/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/DeltaPlugin.scala
@@ -80,7 +80,7 @@ class DeltaPlugin(
     val path = Try {
       val targetFileIndex = extractValue[AnyRef](command, fieldName)
       extractValue[org.apache.hadoop.fs.Path](targetFileIndex, "path")
-    }.getOrElse {
+    } getOrElse {
       val deltaLog = extractValue[AnyRef](command, "deltaLog")
       extractValue[AnyRef](deltaLog, "dataPath")
     }

--- a/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
+++ b/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
@@ -1,0 +1,63 @@
+/*
+ * Copyright 2021 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.issue
+
+import za.co.absa.commons.io.TempDirectory
+import za.co.absa.spline.SparkApp
+
+/**
+ * This Job requires Spark 3 or higher
+ */
+object DeltaMergeDSV2Job extends SparkApp(
+  name = "DeltaMergeDSV2Job",
+  conf = Seq(
+    ("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension"),
+    ("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog"))
+) {
+
+
+
+  val path = TempDirectory().deleteOnExit().path
+
+  import za.co.absa.spline.harvester.SparkLineageInitializer._
+
+  // Initializing library to hook up to Apache Spark
+  spark.enableLineageTracking()
+
+  spark.sql(s"CREATE DATABASE dsv2 LOCATION '$path'")
+
+  spark.sql("CREATE TABLE dsv2.foo ( ID int, NAME string ) USING DELTA")
+  spark.sql("INSERT INTO dsv2.foo VALUES (1014, 'Warsaw'), (1002, 'Corte')")
+
+
+  spark.sql("CREATE TABLE dsv2.fooUpdate ( ID Int, NAME String ) USING DELTA")
+  spark.sql("INSERT INTO dsv2.fooUpdate VALUES (1014, 'Lodz'), (1003, 'Prague')")
+
+  spark.sql(
+    """
+      | MERGE INTO dsv2.foo
+      | USING dsv2.fooUpdate
+      | ON dsv2.foo.ID = dsv2.fooUpdate.ID
+      | WHEN MATCHED THEN
+      |   UPDATE SET
+      |     NAME = dsv2.fooUpdate.NAME
+      | WHEN NOT MATCHED
+      |  THEN INSERT (ID, NAME)
+      |  VALUES (dsv2.fooUpdate.ID, dsv2.fooUpdate.NAME)
+      |""".stripMargin
+  )
+}

--- a/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
+++ b/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
@@ -28,9 +28,6 @@ object DeltaMergeDSV2Job extends SparkApp(
     ("spark.sql.extensions", "io.delta.sql.DeltaSparkSessionExtension"),
     ("spark.sql.catalog.spark_catalog", "org.apache.spark.sql.delta.catalog.DeltaCatalog"))
 ) {
-
-
-
   val path = TempDirectory().deleteOnExit().path
 
   import za.co.absa.spline.harvester.SparkLineageInitializer._
@@ -43,13 +40,13 @@ object DeltaMergeDSV2Job extends SparkApp(
   spark.sql("CREATE TABLE dsv2.foo ( ID int, NAME string ) USING DELTA")
   spark.sql("INSERT INTO dsv2.foo VALUES (1014, 'Warsaw'), (1002, 'Corte')")
 
-
   spark.sql("CREATE TABLE dsv2.fooUpdate ( ID Int, NAME String ) USING DELTA")
   spark.sql("INSERT INTO dsv2.fooUpdate VALUES (1014, 'Lodz'), (1003, 'Prague')")
 
   spark.sql("CREATE TABLE dsv2.barUpdate ( ID Int, NAME String ) USING DELTA")
   spark.sql("INSERT INTO dsv2.barUpdate VALUES (4242, 'Paris'), (3342, 'Bordeaux')")
 
+  spark.sql(s"UPDATE dsv2.foo SET NAME = 'Korok' WHERE ID == 1002")
 
   spark.sql(
     """
@@ -73,4 +70,6 @@ object DeltaMergeDSV2Job extends SparkApp(
       |  VALUES (foobar.ID, foobar.NAME)
       |""".stripMargin
   )
+
+  spark.sql(s"DELETE FROM dsv2.foo WHERE ID == 1014")
 }

--- a/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
+++ b/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
@@ -47,17 +47,30 @@ object DeltaMergeDSV2Job extends SparkApp(
   spark.sql("CREATE TABLE dsv2.fooUpdate ( ID Int, NAME String ) USING DELTA")
   spark.sql("INSERT INTO dsv2.fooUpdate VALUES (1014, 'Lodz'), (1003, 'Prague')")
 
+  spark.sql("CREATE TABLE dsv2.barUpdate ( ID Int, NAME String ) USING DELTA")
+  spark.sql("INSERT INTO dsv2.barUpdate VALUES (4242, 'Paris'), (3342, 'Bordeaux')")
+
+
+  spark.sql(
+    """
+      | CREATE OR REPLACE VIEW tempview AS
+      |   SELECT * FROM dsv2.fooUpdate
+      |   UNION
+      |   SELECT * FROM dsv2.barUpdate
+      |""".stripMargin
+  )
+
   spark.sql(
     """
       | MERGE INTO dsv2.foo
-      | USING dsv2.fooUpdate
-      | ON dsv2.foo.ID = dsv2.fooUpdate.ID
+      | USING tempview AS foobar
+      | ON dsv2.foo.ID = foobar.ID
       | WHEN MATCHED THEN
       |   UPDATE SET
-      |     NAME = dsv2.fooUpdate.NAME
+      |     NAME = foobar.NAME
       | WHEN NOT MATCHED
       |  THEN INSERT (ID, NAME)
-      |  VALUES (dsv2.fooUpdate.ID, dsv2.fooUpdate.NAME)
+      |  VALUES (foobar.ID, foobar.NAME)
       |""".stripMargin
   )
 }

--- a/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
+++ b/examples/src/main/scala/za/co/absa/spline/issue/DeltaMergeDSV2Job.scala
@@ -35,7 +35,7 @@ object DeltaMergeDSV2Job extends SparkApp(
   // Initializing library to hook up to Apache Spark
   spark.enableLineageTracking()
 
-  spark.sql(s"CREATE DATABASE dsv2 LOCATION '$path'")
+  spark.sql("CREATE DATABASE dsv2 LOCATION '$path'")
 
   spark.sql("CREATE TABLE dsv2.foo ( ID int, NAME string ) USING DELTA")
   spark.sql("INSERT INTO dsv2.foo VALUES (1014, 'Warsaw'), (1002, 'Corte')")
@@ -46,7 +46,7 @@ object DeltaMergeDSV2Job extends SparkApp(
   spark.sql("CREATE TABLE dsv2.barUpdate ( ID Int, NAME String ) USING DELTA")
   spark.sql("INSERT INTO dsv2.barUpdate VALUES (4242, 'Paris'), (3342, 'Bordeaux')")
 
-  spark.sql(s"UPDATE dsv2.foo SET NAME = 'Korok' WHERE ID == 1002")
+  spark.sql("UPDATE dsv2.foo SET NAME = 'Korok' WHERE ID == 1002")
 
   spark.sql(
     """
@@ -71,5 +71,5 @@ object DeltaMergeDSV2Job extends SparkApp(
       |""".stripMargin
   )
 
-  spark.sql(s"DELETE FROM dsv2.foo WHERE ID == 1014")
+  spark.sql("DELETE FROM dsv2.foo WHERE ID == 1014")
 }

--- a/integration-tests/src/test/scala/za/co/absa/spline/DeltaDSV2Spec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/DeltaDSV2Spec.scala
@@ -53,7 +53,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
 
           for {
             (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
-              spark.sql(s"CREATE TABLE foo (ID int, NAME string) USING delta")
+              spark.sql("CREATE TABLE foo (ID int, NAME string) USING delta")
               testData.write.format("delta").mode("append").saveAsTable("foo")
             }
           } yield {
@@ -82,7 +82,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
 
           for {
             (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
-              spark.sql(s"CREATE TABLE foo (ID int, NAME string) USING delta")
+              spark.sql("CREATE TABLE foo (ID int, NAME string) USING delta")
               testData.write.format("delta").mode("overwrite").insertInto("foo")
             }
           } yield {
@@ -115,7 +115,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
 
           for {
             (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
-              spark.sql(s"CREATE TABLE foo (ID int, NAME string) USING delta PARTITIONED BY (ID)")
+              spark.sql("CREATE TABLE foo (ID int, NAME string) USING delta PARTITIONED BY (ID)")
               spark.sql(
                 """
                   |INSERT OVERWRITE foo PARTITION (ID = 222222)
@@ -157,7 +157,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
 
           for {
             (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
-              spark.sql(s"CREATE TABLE foo (ID int, NAME string) USING delta PARTITIONED BY (NAME)")
+              spark.sql("CREATE TABLE foo (ID int, NAME string) USING delta PARTITIONED BY (NAME)")
               spark.sql(
                 """
                   |INSERT OVERWRITE foo PARTITION (NAME)
@@ -218,7 +218,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
 
           for {
             (plan1, Seq(event1)) <- lineageCaptor.lineageOf {
-              spark.sql(s"CREATE TABLE foo (toBeOrNotToBe boolean) USING delta")
+              spark.sql("CREATE TABLE foo (toBeOrNotToBe boolean) USING delta")
               testData.write.format("delta").mode("overwrite").option("overwriteSchema", "true").saveAsTable("foo")
             }
           } yield {
@@ -250,7 +250,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
               testData.write.format("delta").saveAsTable("foo")
             }
             (plan2, Seq(event2)) <- lineageCaptor.lineageOf {
-              spark.sql(s"DELETE FROM foo WHERE ID == 1014")
+              spark.sql("DELETE FROM foo WHERE ID == 1014")
             }
           } yield {
             plan2.id.value shouldEqual event2.planId
@@ -283,7 +283,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
               testData.write.format("delta").saveAsTable("foo")
             }
             (plan2, Seq(event2)) <- lineageCaptor.lineageOf {
-              spark.sql(s"UPDATE foo SET NAME = 'Korok' WHERE ID == 1002")
+              spark.sql("UPDATE foo SET NAME = 'Korok' WHERE ID == 1002")
             }
           } yield {
             plan2.id.value shouldEqual event2.planId

--- a/integration-tests/src/test/scala/za/co/absa/spline/DeltaDSV2Spec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/DeltaDSV2Spec.scala
@@ -307,34 +307,40 @@ class DeltaDSV2Spec extends AsyncFlatSpec
     ) { implicit spark =>
       withLineageTracking { lineageCaptor =>
         withDatabase("testDB") {
-          val testData = {
-            import spark.implicits._
-            Seq((1014, "Warsaw"), (1002, "Corte")).toDF("ID", "NAME")
-          }
-          val updateData = {
-            import spark.implicits._
-            Seq((1014, "Lodz"), (1003, "Prague")).toDF("ID", "NAME")
-          }
-
           for {
             (_, _) <- lineageCaptor.lineageOf {
-              testData.write.format("delta").saveAsTable("foo")
+              spark.sql("CREATE TABLE foo ( ID int, NAME string ) USING DELTA")
+              spark.sql("INSERT INTO foo VALUES (1014, 'Warsaw'), (1002, 'Corte')")
             }
             (_, _) <- lineageCaptor.lineageOf {
-              updateData.write.format("delta").saveAsTable("fooUpdate")
+              spark.sql("CREATE TABLE fooUpdate ( ID Int, NAME String ) USING DELTA")
+              spark.sql("INSERT INTO fooUpdate VALUES (1014, 'Lodz'), (1003, 'Prague')")
+            }
+            (_, _) <- lineageCaptor.lineageOf {
+              spark.sql("CREATE TABLE barUpdate ( ID Int, NAME String ) USING DELTA")
+              spark.sql("INSERT INTO barUpdate VALUES (4242, 'Paris'), (3342, 'Bordeaux')")
             }
             (plan, Seq(event)) <- lineageCaptor.lineageOf {
               spark.sql(
                 """
+                  | CREATE OR REPLACE VIEW tempview AS
+                  |   SELECT * FROM fooUpdate
+                  |   UNION
+                  |   SELECT * FROM barUpdate
+                  |""".stripMargin
+              )
+
+              spark.sql(
+                """
                   | MERGE INTO foo
-                  | USING fooUpdate
-                  | ON foo.ID = fooUpdate.ID
+                  | USING tempview AS foobar
+                  | ON foo.ID = foobar.ID
                   | WHEN MATCHED THEN
                   |   UPDATE SET
-                  |     NAME = fooUpdate.NAME
+                  |     NAME = foobar.NAME
                   | WHEN NOT MATCHED
                   |  THEN INSERT (ID, NAME)
-                  |  VALUES (fooUpdate.ID, fooUpdate.NAME)
+                  |  VALUES (foobar.ID, foobar.NAME)
                   |""".stripMargin
               )
             }
@@ -349,16 +355,20 @@ class DeltaDSV2Spec extends AsyncFlatSpec
             val mergeOp = plan.operations.write.childOperation
             mergeOp.params("condition").asInstanceOf[Option[String]].value should include("ID")
 
-            val reads = mergeOp.childOperations.map(_.asInstanceOf[ReadOperation])
+            val reads =  plan.operations.reads
 
             val mergeOutput = mergeOp.outputAttributes
             val read0Output = reads(0).outputAttributes
             val read1Output = reads(1).outputAttributes
+            val read2Output = reads(2).outputAttributes
 
             mergeOutput(0) should dependOn(read0Output(0))
             mergeOutput(1) should dependOn(read0Output(1))
             mergeOutput(0) should dependOn(read1Output(0))
             mergeOutput(1) should dependOn(read1Output(1))
+            mergeOutput(0) should dependOn(read2Output(0))
+            mergeOutput(1) should dependOn(read2Output(1))
+
           }
         }
       }

--- a/integration-tests/src/test/scala/za/co/absa/spline/DeltaDSV2Spec.scala
+++ b/integration-tests/src/test/scala/za/co/absa/spline/DeltaDSV2Spec.scala
@@ -353,7 +353,7 @@ class DeltaDSV2Spec extends AsyncFlatSpec
             plan.operations.write.outputSource should endWith("/testdb.db/foo")
 
             val mergeOp = plan.operations.write.childOperation
-            mergeOp.params("condition").asInstanceOf[Option[String]].value should include("ID")
+            mergeOp.params("condition").asInstanceOf[String] should include("ID")
 
             val reads =  plan.operations.reads
 

